### PR TITLE
Feat/djr model abstraction

### PR DESCRIPTION
**Pull request by DarylRodrigo**

### CODE REVIEW

File: PPO/Config.py [hunk 1]
```diff
@@ -10,18 +10,17 @@ def __init__(self, env):
...
 
     self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    print("Running experiment with device: {}".format(self.device))
     self.seed = 123456789
 
     self.n_episodes = 2000
     self.max_t = 1000
     # TODO set as decaying and pass into learn from PPO
-    self.epsilon = 0.2
+    self.epsilon = 0.1
```
Comment (instance01) [inline PPO/Config.py:19]: Nice. I'm pretty excited about whether this will work.

File: PPO/Networks.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import torch.nn as nn
+import numpy as np
+import torch
+from Config import Config
+import pdb
+
+# TODO:
+# - Layer initialisation
```
Comment (instance01) [inline PPO/Networks.py:8]: I suppose you want to make this completely configurable correct?

File: PPO/Networks.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      nn.ReLU(),
+      nn.Flatten(),
+      nn.Linear(64 * 7 * 7, self.config.hidden_size),
+      nn.ReLU(),
+    )
+
+  def forward(self, state):
+    return self.head(state)
+
+def head_model(config: Config) -> nn.Sequential:
```
Comment (instance01) [inline PPO/Networks.py:33]: Moving out of Models was a good idea.

File: Policy Gradient/src/PPO/PPO.py [hunk 4]
```diff
@@ -115,7 +170,7 @@ def learn(self, num_learn):
       # find ratios
       actions, log_probs, values, entropy = self.model.evaluate(prev_states, prev_actions)
       ratio = torch.exp(log_probs - prev_log_probs.detach())
-
+
```
Comment (instance01) [inline Policy Gradient/src/PPO/PPO.py:173]: just a nitpick - unnecessary whitespaces :D

File: PPO/PPO.py [hunk 5]
```diff
@@ -1,10 +1,12 @@
 import torch
 import torch.optim as optim
 import torch.nn.functional as F
+import torch.nn as nn
 import copy
 import pdb
+import numpy as np
 
-class PPO:
+class PPOBase:
```
Comment (instance01) [inline PPO/PPO.py:9]: Nice idea.

File: PPO/util.py [hunk 6]
```diff
@@ -25,31 +26,87 @@ def train(config):
...
     
     if np.mean(scores_deque) > config.win_condition:
-      print("\rEnvironment Solved!")
+      print("\nEnvironment Solved!")
       break
 
-  return scores
+  return scores, average_scores
+
+def train_pixel(config):
```
Comment (instance01) [inline PPO/util.py:60]: Should probably move this to somewhere else and not util, seems like a heavy util function. It's more like a main.

What I usually do is an Experiment class that encapsulates the current config and provides a train and evaluation function. Something we could consider.

File: PPO/util.py [hunk 7]
```diff
@@ -25,31 +26,87 @@ def train(config):
...
+  env = copy.deepcopy(config.env)
```
Comment (instance01) [inline PPO/util.py:61]: I'm actually against a deepcopy again, for instance the box2d envs don't support that etc., we shouldn't use it.

File: Policy Gradient/src/PPO/PPO.py [hunk 8]
```diff
@@ -68,9 +67,63 @@ def learn(self, num_learn):
       self.optimiser.zero_grad()
       loss.backward()
       self.optimiser.step()
-
+    
     self.model_old.load_state_dict(self.model.state_dict())
+  
+  def learn_gae(self, num_learn, last_value, last_done):
```
Comment (instance01) [inline Policy Gradient/src/PPO/PPO.py:73]: Very nice, but wrong file I think? This is under the `Policy Gradient` folder.

File: PPO/PPO.py [hunk 9]
```diff
@@ -67,3 +87,85 @@ def learn(self, num_learn):
...
+        nextnonterminal = 1.0 - self.mem.dones[t+1]
+        next_return = discounted_returns[t+1]
+      discounted_returns[t] = self.mem.rewards[t] + self.gamma * nextnonterminal * next_return
+
+    # normalise rewards
+    discounted_returns = torch.FloatTensor(discounted_returns).to(self.device)
+    discounted_returns = (discounted_returns - discounted_returns.mean()) / (discounted_returns.std() + 1e-8)
+    
+    # Fetch collected state, action and log_probs from memory & reshape for nn
+    prev_states = torch.stack(self.mem.states).reshape((-1,)+(4, 84, 84)).to(self.device).detach()
```
Comment (instance01) [inline PPO/PPO.py:131]: Currently hardcoded shape (e.g. won't work with procgen)

File: PPO/PPO.py [hunk 10]
```diff
@@ -67,3 +87,85 @@ def learn(self, num_learn):
...
+      actions, log_probs, values, entropy = self.model.act(prev_states, prev_actions)
+      ratio = torch.exp(log_probs - prev_log_probs.detach())
+      values = values.squeeze() * 0.5
+
+      # Stats
+      approx_kl = (prev_log_probs - log_probs).mean()
+
+      # calculate advantage & normalise
+      advantage = discounted_returns - values
+      advantage = (advantage - advantage.mean()) / (advantage.std() + 1e-8)
```
Comment (instance01) [inline PPO/PPO.py:147]: `1e-8` - excellent stuff

**Review by instance01 — APPROVED** [review]

File: PPO/Networks.py (hunk 2 above)
Comment (DarylRodrigo) [reply PPO/Networks.py:8]: I'm actually not too sure, we can have the different networks with different initialisations, I don't really see a scenario where we wouldn't want to have it though WDYT?

File: Policy Gradient/src/PPO/PPO.py (hunk 8 above)
Comment (DarylRodrigo) [reply Policy Gradient/src/PPO/PPO.py:73]: Yeh this isn't actually working I think :P

File: PPO/PPO.py (hunk 9 above)
Comment (DarylRodrigo) [reply PPO/PPO.py:131]: yeh definitely, the env that I used right now comes out as something weird and I couldn't use it. We can amend this when we do the multi env.

File: PPO/PPO.py (hunk 10 above)
Comment (DarylRodrigo) [reply PPO/PPO.py:147]: noice, we should definitely make this configurable swell
